include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -184,56 +184,51 @@ struct __init_processing
 // transform_reduce
 //------------------------------------------------------------------------
 
-template <typename _ExecutionPolicy, typename _Operation1, typename _Operation2, typename _Tp>
-struct transform_reduce_seq
+template <typename _ExecutionPolicy, typename _Tp, typename _Operation1, typename _Operation2,
+          ::std::size_t __iters_per_work_item = 0>
+struct transform_reduce_known
 {
+  private:
+    // Apply unary and binary operations __n values and return result.
+    template <typename _BinaryOp, typename _UnaryOp, typename... _Acc>
+    static _Tp
+    apply_unary_binary(const ::std::size_t __items_to_process, const ::std::size_t __adjusted_global_id,
+                       _BinaryOp __binary_op, _UnaryOp __unary_op, const _Acc&... __acc)
+    {
+        // Keep these statements in the same scope to allow for better memory alignment.
+        // Vectorized access if __items_to_process is known at compile time.
+        _Tp __res = __unary_op(__adjusted_global_id, __acc...);
+        for (::std::size_t __i = 1; __i < __items_to_process; ++__i)
+            __res = __binary_op(__res, __unary_op(__adjusted_global_id + __i, __acc...));
+        return __res;
+    }
```

Comment:
I'm probably missing something, but how can we take advantage of compile time knowledge of `__items_to_process` when it is a runtime variable in this function.  We also lose the `#pragma unroll` in this commit.  Have we confirmed that performance isn't impacted by this refactor?

If we want to unify code, but keep the compile time knowledge of the for loop in the external scope, we can unify the 
body of the for loop with a similar structure.  However, I'm not sure if it offers much benefit to replace the single line statement with a single line function call.

Reply:
In case `__items_to_process` is the full set of `__iters_per_work_item` items, the template parameter `__iters_per_work_item` is passed to `apply_unary_binary` for its `__items_to_process` argument. Thus, the length of the for loop should be known at compile time and the compiler can unroll and vectorize the loop. The performance is unchanged on the platforms tested.

Reply:
Interesting, I feel like we are trusting the compiler to do a lot for us to achieve good performance here but it seems to be working out.  I wonder with different compilers or small seemingly innocuous future changes to the code if we may have an unexpected degradation in performance.  

@akukanov, @MikeDvorskiy or any others: Do you think that we can be confident that any compiler should be able to figure this out and generate performant code? (generating a vectorized and aligned version as well as a dynamic version of this function)
If not, are there any strategies you can recommend here to keep this unification of redundant code, but give better clues to the compiler to ensure it generates code we are happy with?

Reply:
As for me, it is difficult to say about  that.... I had not such experiments... But I would expect that "yes" in the "ideal world" :) Probably, it should be discussed with a guy from the compiler team or with a person who has expertise here - may be Ma, Zhiqiang <zhiqiang.ma@intel.com>.